File: backend/internal/export/xlsx.go
package export

import (
	"archive/zip"
	"bytes"
	"encoding/xml"
	"fmt"
	"io"
	"strconv"
	"strings"
)

// Sheet is the intentionally small export seam: domain code supplies stable
// headers and rows, while this package owns the XLSX container and XML rules.
type Sheet struct {
	Name    string
	Headers []string
	Rows    [][]string
}

func Build(sheets []Sheet, version string) ([]byte, error) {
	var output bytes.Buffer
	archive := zip.NewWriter(&output)
	files := map[string]string{
		"[Content_Types].xml":        contentTypesXML(len(sheets)),
		"_rels/.rels":                relsXML(),
		"xl/workbook.xml":            workbookXML(sheets),
		"xl/_rels/workbook.xml.rels": workbookRelsXML(len(sheets)),
		"xl/styles.xml":              stylesXML(),
		"docProps/core.xml":          coreXML(version),
		"docProps/app.xml":           appXML(len(sheets)),
	}
	for name, contents := range files {
		writer, err := archive.Create(name)
		if err != nil {
			return nil, err
		}
		if _, err := io.WriteString(writer, contents); err != nil {
			return nil, err
		}
	}
	for index, sheet := range sheets {
		writer, err := archive.Create(fmt.Sprintf("xl/worksheets/sheet%d.xml", index+1))
		if err != nil {
			return nil, err
		}
		if _, err := io.WriteString(writer, sheetXML(sheet)); err != nil {
			return nil, err
		}
	}
	if err := archive.Close(); err != nil {
		return nil, err
	}
	return output.Bytes(), nil
}

func xmlText(value string) string {
	encoded, _ := xml.Marshal(value)
	text := string(encoded)
	return strings.TrimSuffix(strings.TrimPrefix(text, "<string>"), "</string>")
}
func contentTypesXML(count int) string {
	parts := []string{`<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types"><Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/><Default Extension="xml" ContentType="application/xml"/><Override PartName="/xl/workbook.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/><Override PartName="/xl/styles.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.styles+xml"/>`}
	for index := 1; index <= count; index++ {
		parts = append(parts, fmt.Sprintf(`<Override PartName="/xl/worksheets/sheet%d.xml" ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/>`, index))
	}
	return strings.Join(parts, "") + `</Types>`
}
func relsXML() string {
	return `<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships"><Relationship Id="rId1" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" Target="xl/workbook.xml"/><Relationship Id="rId2" Type="http://schemas.openxmlformats.org/package/2006/relationships/metadata/core-properties" Target="docProps/core.xml"/><Relationship Id="rId3" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/extended-properties" Target="docProps/app.xml"/></Relationships>`
}
func workbookXML(sheets []Sheet) string {
	var b strings.Builder
	b.WriteString(`<?xml version="1.0" encoding="UTF-8" standalone="yes"?><workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships"><sheets>`)
	for index, sheet := range sheets {
		fmt.Fprintf(&b, `<sheet name="%s" sheetId="%d" r:id="rId%d"/>`, xmlText(sheet.Name), index+1, index+1)
	}
	b.WriteString(`</sheets></workbook>`)
	return b.String()
}
func workbookRelsXML(count int) string {
	var b strings.Builder
	b.WriteString(`<?xml version="1.0" encoding="UTF-8" standalone="yes"?><Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">`)
	for index := 1; index <= count; index++ {
		fmt.Fprintf(&b, `<Relationship Id="rId%d" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" Target="worksheets/sheet%d.xml"/>`, index, index)
	}
	b.WriteString(`<Relationship Id="rId50" Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/styles" Target="styles.xml"/></Relationships>`)
	return b.String()
}
func stylesXML() string {
	return `<?xml version="1.0" encoding="UTF-8"?><styleSheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><fonts count="1"><font><sz val="11"/><name val="Calibri"/></font></fonts><fills count="2"><fill><patternFill patternType="none"/></fill><fill><patternFill patternType="gray125"/></fill></fills><borders count="1"><border/></borders><cellStyleXfs count="1"><xf numFmtId="0" fontId="0" fillId="0" borderId="0"/></cellStyleXfs><cellXfs count="2"><xf numFmtId="0" fontId="0" fillId="0" borderId="0"/><xf numFmtId="0" fontId="0" fillId="0" borderId="0" applyFont="1"/></cellXfs></styleSheet>`
}
func coreXML(version string) string {
	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?><cp:coreProperties xmlns:cp="http://purl.org/dc/elements/1.1/"><dc:title xmlns:dc="http://purl.org/dc/elements/1.1/">ChronoFish export</dc:title><dc:creator xmlns:dc="http://purl.org/dc/elements/1.1/">ChronoFish</dc:creator><dc:description xmlns:dc="http://purl.org/dc/elements/1.1/">%s</dc:description></cp:coreProperties>`, xmlText(version))
}
func appXML(count int) string {
	return fmt.Sprintf(`<?xml version="1.0" encoding="UTF-8"?><Properties xmlns="http://schemas.openxmlformats.org/officeDocument/2006/extended-properties"><Application>ChronoFish</Application><Sheets>%d</Sheets></Properties>`, count)
}
func sheetXML(sheet Sheet) string {
	var b strings.Builder
	b.WriteString(`<?xml version="1.0" encoding="UTF-8" standalone="yes"?><worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><sheetData>`)
	writeRowXML(&b, 1, sheet.Headers, true)
	for index, row := range sheet.Rows {
		writeRowXML(&b, index+2, row, false)
	}
	b.WriteString(`</sheetData></worksheet>`)
	return b.String()
}
func writeRowXML(b *strings.Builder, rowNumber int, values []string, header bool) {
	fmt.Fprintf(b, `<row r="%d">`, rowNumber)
	for index, value := range values {
		cell := columnName(index+1) + strconv.Itoa(rowNumber)
		style := ""
		if header {
			style = ` s="1"`
		}
		if !header && value != "" {
			if _, err := strconv.ParseFloat(value, 64); err == nil {
				fmt.Fprintf(b, `<c r="%s"%s><v>%s</v></c>`, cell, style, xmlText(value))
				continue
			}
		}
		fmt.Fprintf(b, `<c r="%s" t="inlineStr"%s><is><t>%s</t></is></c>`, cell, style, xmlText(value))
	}
	b.WriteString(`</row>`)
}
func columnName(number int) string {
	result := ""
	for number > 0 {
		number--
		result = string(rune('A'+number%26)) + result
		number /= 26
	}
	return result
}

File: backend/src/chronofish/reporting/xlsx.py
from __future__ import annotations

import re
from io import BytesIO
from numbers import Number
from xml.sax.saxutils import escape, quoteattr
from zipfile import ZIP_DEFLATED, ZipFile

Sheet = tuple[str, list[str], list[list[object]]]
INVALID_XML_CHARACTERS = re.compile(r"[^\x09\x0a\x0d\x20-\ud7ff\ue000-\ufffd\U00010000-\U0010ffff]")


def _column(number: int) -> str:
    result = ""
    while number:
        number, remainder = divmod(number - 1, 26)
        result = chr(65 + remainder) + result
    return result


def _cell_xml(reference: str, raw: object) -> str:
    if isinstance(raw, bool):
        return f'<c r="{reference}" t="b"><v>{int(raw)}</v></c>'
    if isinstance(raw, Number):
        return f'<c r="{reference}"><v>{raw}</v></c>'
    value = "" if raw is None else INVALID_XML_CHARACTERS.sub("", str(raw))
    return f'<c r="{reference}" t="inlineStr"><is><t xml:space="preserve">{escape(value)}</t></is></c>'


def _row_xml(number: int, values: list[object]) -> str:
    cells = [_cell_xml(f"{_column(index)}{number}", raw) for index, raw in enumerate(values, 1)]
    return f'<row r="{number}">{"".join(cells)}</row>'


def _sheet_xml(headers: list[str], rows: list[list[object]]) -> str:
    body = [_row_xml(1, headers)]
    body.extend(_row_xml(index, row) for index, row in enumerate(rows, 2))
    return (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<worksheet xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main">'
        f"<sheetData>{''.join(body)}</sheetData></worksheet>"
    )


def build_xlsx(sheets: list[Sheet]) -> bytes:
    content_types = [
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>',
        '<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">',
        '<Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml"/>',
        '<Default Extension="xml" ContentType="application/xml"/>',
        '<Override PartName="/xl/workbook.xml" '
        'ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet.main+xml"/>',
    ]
    content_types.extend(
        f'<Override PartName="/xl/worksheets/sheet{index}.xml" '
        'ContentType="application/vnd.openxmlformats-officedocument.spreadsheetml.worksheet+xml"/>'
        for index in range(1, len(sheets) + 1)
    )
    content_types.append("</Types>")
    workbook_sheets = "".join(
        f'<sheet name={quoteattr(name)} sheetId="{index}" r:id="rId{index}"/>'
        for index, (name, _headers, _rows) in enumerate(sheets, 1)
    )
    workbook = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<workbook xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main" '
        'xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships">'
        f"<sheets>{workbook_sheets}</sheets></workbook>"
    )
    workbook_rels = "".join(
        f'<Relationship Id="rId{index}" '
        'Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet" '
        f'Target="worksheets/sheet{index}.xml"/>'
        for index in range(1, len(sheets) + 1)
    )
    root_rels = (
        '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
        '<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">'
        '<Relationship Id="rId1" '
        'Type="http://schemas.openxmlformats.org/officeDocument/2006/relationships/officeDocument" '
        'Target="xl/workbook.xml"/>'
        "</Relationships>"
    )
    output = BytesIO()
    with ZipFile(output, "w", ZIP_DEFLATED) as archive:
        archive.writestr("[Content_Types].xml", "".join(content_types))
        archive.writestr("_rels/.rels", root_rels)
        archive.writestr("xl/workbook.xml", workbook)
        archive.writestr(
            "xl/_rels/workbook.xml.rels",
            '<?xml version="1.0" encoding="UTF-8" standalone="yes"?>'
            '<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">'
            f"{workbook_rels}</Relationships>",
        )
        for index, (_name, headers, rows) in enumerate(sheets, 1):
            archive.writestr(f"xl/worksheets/sheet{index}.xml", _sheet_xml(headers, rows))
    return output.getvalue()
